xrtpy/response/effective_area.py

```diff
@@ -0,0 +1,478 @@
+__all__ = [
+    "EffectiveAreaFundamental",
+    "effective_area",
+]
+
+import math
+import numpy as np
+import pkg_resources
+import scipy.io
+import sunpy.io.special
+import sunpy.time
+
+from astropy import units as u
+from datetime import timedelta
+from functools import cached_property
+from scipy import interpolate
+
+from xrtpy.response.channel import Channel, resolve_filter_name
+from xrtpy.util.time import epoch
+
+index_mapping_to_fw1_name = {
+    "Open": 0,
+    "Al-poly": 1,
+    "C-poly": 2,
+    "Be-thin": 3,
+    "Be-med": 4,
+    "Al-med": 5,
+}
+
+index_mapping_to_fw2_name = {
+    "Open": 0,
+    "Al-mesh": 1,
+    "Ti-poly": 2,
+    "G-band": 3,
+    "Al-thick": 4,
+    "Be-thick": 5,
+}
+
+
+_ccd_contam_filename = pkg_resources.resource_filename(
+    "xrtpy", "data/channels/xrt_contam_on_ccd.geny"
+)
+_filter_contam_filename = pkg_resources.resource_filename(
+    "xrtpy", "data/channels/xrt_contam_on_filter.geny"
+)
+
+_ccd_contam_file = scipy.io.readsav(_ccd_contam_filename)
+_filter_contam_file = scipy.io.readsav(_filter_contam_filename)
+
+# CCD contam geny files keys for time and date.
+_ccd_contamination_file_time = _ccd_contam_file["p1"]
+_ccd_contamination = _ccd_contam_file["p2"]
+
+# Filter contam geny files keys for time and date.
+_filter_contamination_file_time = _filter_contam_file["p1"]
+_filter_contamination = _filter_contam_file["p2"]
+
+
+class EffectiveAreaFundamental:
+    """
+    Class for calculating the effective area.
+
+    Parameters
+    -----------
+    filter_name : str
+        The name of the filter.
+
+    observation_date: str date-time
+        The date of the observation.  For valid date formats, look at the documentation for
+        `sunpy.time.parse_time`.
+    """
+
+    def __init__(self, filter_name, observation_date):
+        self._name = resolve_filter_name(filter_name)
+        self.observation_date = observation_date
+        self._channel = Channel(self.name)
+
+    @property
+    def name(self) -> str:
+        """Name of XRT X-Ray channel filter."""
+        return self._name
+
+    @property
+    def observation_date(self) -> str:
+        """Date of observation."""
+        return self._observation_date
+
+    @observation_date.setter
+    def observation_date(self, date):
+        """Validating users requested observation date."""
+        astropy_time = sunpy.time.parse_time(date)  # Astropy time in utc
+        observation_date = astropy_time.datetime
+
+        if observation_date <= epoch:
+            raise ValueError(
+                "Invalid date: {:}.\n Date must be after September 22nd, 2006 21:36:00.".format(
+                    observation_date
```

Comment:
This could be an [f-string](https://realpython.com/python-f-strings/#f-strings-a-new-and-improved-way-to-format-strings-in-python) too.